python/ray/_private/runtime_env/packaging.py

```diff
@@ -409,19 +447,102 @@ def download_and_unpack_package(
                 with open(pkg_uri, "rb", transport_params=tp) as package_zip:
                     with open(pkg_file, "wb") as fin:
                         fin.write(package_zip.read())
+                unzip_package(pkg_file, local_dir, True, True, logger)
+            elif protocol == Protocol.GS or protocol == Protocol.HTTPS:
+                # Download package via GS or HTTPS.
+                try:
+                    from smart_open import open
+                except ImportError:
+                    raise ImportError(
+                        "You must `pip install smart_open` "
+                        f"to fetch {protocol.value.upper()} URIs.")
+                with open(pkg_uri, "rb") as package_zip:
+                    with open(pkg_file, "wb") as fin:
+                        fin.write(package_zip.read())
+                unzip_package(pkg_file, local_dir, True, True, logger)
             else:
                 raise NotImplementedError(
                     f"Protocol {protocol} is not supported")
 
-            os.mkdir(local_dir)
-            logger.debug(f"Unpacking {pkg_file} to {local_dir}")
-            with ZipFile(str(pkg_file), "r") as zip_ref:
-                zip_ref.extractall(local_dir)
-            pkg_file.unlink()
-
         return str(local_dir)
 
 
+def get_top_level_dir_from_compressed_package(package_path: str):
```

Comment:
unit test ? this function is quite self-contained and directly related to top-level directory handling

Reply:
I added the `TestGetTopLevelDirFromCompressedPackage` test class in `test_runtime_env_packaging.py` to cover this function's logic.